crmd/membership.c

```diff
@@ -438,12 +438,24 @@ crm_update_quorum(gboolean quorum, gboolean force_update)
         fsa_register_cib_callback(call_id, FALSE, NULL, cib_quorum_update_complete);
         free_xml(update);
 
-        /* If a node not running any resources is cleanly shut down and drops us
-         * below quorum, we won't necessarily abort the transition, so abort it
-         * here to be safe.
+        /* Quorum changes usually cause a new transition via other activity:
+         * quorum gained via a node joining will abort via the node join,
+         * and quorum lost via a node leaving will usually abort via resource
+         * activity and/or fencing.
+         *
+         * However, it is possible that nothing else causes a transition (e.g.
+         * someone forces quorum via corosync-cmaptcl, or quorum is lost due to
+         * a node in standby shutting down cleanly), so here ensure a new
+         * transition is triggered.
          */
-        if (quorum == FALSE) {
-            abort_transition(INFINITY, tg_restart, "Quorum loss", NULL);
+        if (quorum) {
+            /* If quorum was gained, abort after a short delay, in case multiple
+             * nodes are joining around the same time, so the one that brings us
+             * to quorum doesn't cause all the remaining ones to be fenced.
+             */
+            abort_after_delay(INFINITY, tg_restart, "Quorum gained", 5000);
```

Comment:
If a joining peer makes the cluster acquire the quorum from corosync level, meanwhile if its sbd has `SBD_DELAY_START` enabled, which is usually longer than 5s and postpones start of pacemaker there, unnecessary startup fencing targeting the peer is always triggered: 5s after the quorum has been acquired before the node gets the chance to join at pacemaker level.

I'm trying to think of some potential solutions such as:

- Let pacemaker-controld recognize `SBD_DELAY_START` setting and use a value greater than that as the delay parameter of abort_after_delay() here so that abortion of transition waits for the peer to join CPG. But it's not necessarily straight-forward since `SBD_DELAY_START` can also be just set to `yes`, which makes it adapt the value of `msgwait` in case of disk mode otherwise `2 * watchdog_timeout` ...

- Postpone start of corosync as well by adding the following dependency in sbd.service:

`Before=corosync.service`

But it wouldn't work for the case of diskless SBD, since it apparently requires the connection to corosync to report a successful start:

https://github.com/ClusterLabs/sbd/commit/7ef73ab6a02bbacec147e9f36f29f9a668fb59fd

- Ask users to set this for corosync.service:

`ExecStartPre=/bin/sleep <time>`

The `<time>` is corresponding to `SBD_DELAY_START` configured in /etc/sysconfig/sbd

It'd be bothering for the users since they would have to pay attention to it and make sure  the relevant settings are synchronized in between.

- We probably could let acquiring of "quorum" require the peer to show up at CPG level as well besides at corosync's quorum level? 

Or we probably could only specifically address the case where "wait_for_all" is enabled in corosync and let acquiring of "quorum" require all the peers to show up at CPG level?

So far I cannot think of any better ideas. What do you think @kgaillot and @wenningerk ?

Reply:
Keep in mind that even if we don't abort the transition here, some unrelated event could abort the transition, and we'd still fence the node.

Why are we fencing the node to begin with, if it's in the cluster membership? I wouldn't expect CPG membership to be required.

Reply:
> Keep in mind that even if we don't abort the transition here, some unrelated event could abort the transition, and we'd still fence the node.

Indeed. It's just that the situation easily occurs with this predictable transition abortion.

> 
> Why are we fencing the node to begin with, if it's in the cluster membership? I wouldn't expect CPG membership to be required.

Well, now AFAICS it's related to whether the uname of the pending node is known yet upon creating a node_state entry for it,  and how scheduler considers the status of the node under the situation...

Please take a look if this makes sense when you get a chance: https://github.com/ClusterLabs/pacemaker/pull/3031. Thanks.

Reply:
To make things even more complicated to consider in case we're running corosync in "2-node" (probably same as "wait_for_all" for most of the cases) we're already ignoring quorum and would rather be going for availability of the peer via CPG. At least I guess this is also relevant for startup ... have to check how I did that exactly ...